### ISSUE TITLE
Provide an apt proxy for (cli) crates?

### ISSUE THREAD

**Issue by thomaseizinger** [rust-lang/crates.io#2247]
Rust is pretty popular for writing commandline tools. If you are a Rust developer yourself, you can fairly easily install those by doing `cargo install XYZ`. But unfortunately, this doesn't give you any updates.

I have only been a _user_ of the debian/ubuntu package manager so far and never setup something like this myself but here is an idea:

Would it be possible for crates.io to provide an apt repository that allows users to install commandline tools through apt?

- crates.io should know from the manifest of the of the crate whether it is an application vs a library
- if it makes things easier, we could only support crates that use `cargo-deb` to generate a `.deb` file and provide that

If crates.io cannot or doesn't want to provide such service, are there any webhooks that would allow building such a service and get notified of new uploads happening?

**Comment by jtgeibel** [comment]
Hi @thomaseizinger.  Thanks for the suggestion, however I think this would be a massive amount of scope for the team to try to manage and that we really don't have the skillset necessary to support something like this.

For reference, Debian manages their rust packaging in [this repo](https://salsa.debian.org/rust-team/debcargo-conf/) and while they have some tools that automate parts of their packaging workflow, there is still a lot of manual work that goes into packaging a release.  For instance, the copyright metadata frequently needs tweaks to meet Debian packaging standards.

We also don't have any of the build infrastructure that distributions utilize.  A while back I manually backported a lot of packages to a [PPA](https://launchpad.net/~jtgeibel/+archive/ubuntu/rust-backports) but that required me to sequence through all the layers of dependencies to push things to the build servers in the right order.  This was somewhat difficult even though I was only backporting crates that were already packaged in upstream Debian.  While I think it would be possible to automate some of the backporting work I did here, I think it would be much more difficult to automate the packaging of arbitrary releases.

I'm less familiar with `cargo-deb`.  Maybe it supports creating packages without needing to also package all of the dependencies.  Even if so, a service like this would need to be set up external to core crates.io due to resource (and volunteer effort) constraints.  I think a separate service also makes the most sense from an architecture standpoint.

> are there any webhooks that would allow building such a service and get notified of new uploads happening?

My suggestion for something like this would be to periodically pull down the [index repository](https://github.com/rust-lang/crates.io-index) and look for any new commits.  Commits happen in this repo when versions are published, or yanked/unyanked.  We don't currently have any type of outgoing webhooks to create such event notifications.
